crates/flux-fhir-analysis/src/wf/sortck.rs

```diff
@@ -84,6 +84,34 @@ impl<'genv, 'tcx> InferCtxt<'genv, 'tcx> {
         }
     }
 
+    fn check_constructor(
+        &mut self,
+        expr_span: Span,
+        field_exprs: &[fhir::FieldExpr],
+        _spread: &Option<fhir::Spread>,
+        expected: &rty::Sort,
+    ) -> Result {
+        if let rty::Sort::App(rty::SortCtor::Adt(sort_def), sort_args) = expected {
+            // these are ordered
+            let field_names = sort_def.field_names();
+            let sorts = sort_def.field_sorts(sort_args);
+            for expr in field_exprs {
```

Comment:
So I'm wondering, do we actually need to check that there are no missing fields? 

If someone uses this syntax without specifying all fields, it is simply syntactic sugar for a normal update. That seems like it would be ok?

Reply:
That's only true if you provide a spread, with no spread you have to specify all the fields, e.g., the following is wrong and should give an error


```rust
#[refined_by(x: int, y: int)]
struct S { ... }

#[sig(fn() -> S[S { x: 0 }]
fn foo() -> S { 
    ...
}
``` 


 We've been calling this functional updates, but there's no actual *update* going on here, you are always constructing a new value, by possibly copying some fields from another variable via the spread.

Reply:
@nilehmann I went ahead and added these - I'm not sure if there's a better way to do the diagnostic for missing fields. Lmk if you have any thoughts / objections.

Reply:
Ok - I'm going to work on spreads first and then I'll circle back to this.

Reply:
I think we can simplify the code a bit if we use hashmaps and hash sets instead of vecs. 

I would expose a method in `AdtSortDef` that gives you back an `FxHashMap<Symbol, Sort>` and then

```rust
let mut sort_by_field_name = sort_def.sort_by_field_name(sort_args);
let mut used_fields = FxHashSet::default();
for expr in field_exprs {
    let Some(sort) = sort_by_field_name.get(&expr.ident.name) else {
        // report field not found
    }
    if let Some(old) = used_fields.replace(expr.ident) {
        return Err(...DuplicatefieldUsed...)
    }
    self.check_expr(&expr.expr.sort)?;
}
if let Some(_spred) = spread {
    todo!()
} else if used_fields.len() != sort_by_field_name.len() {
    // report missing fields, we can compute the difference between `sort_by_field_name` and `used_fields` here.
    // I like this approach because we don't clutter the happy path with extra logic needed to remember the 
    // remaining fields.
} else {
   ...
}
```